Composite configuration is not working

Composite configuration is not working because the 

> @ConditionalOnSearchPathLocator 

used in

> org.springframework.cloud.config.server.config.CompositeRepositoryConfiguration.searchPathCompositeEnvironmentRepository 

causes this problem
```
java.lang.IllegalStateException: Error processing condition on org.springframework.cloud.config.server.config.CompositeRepositoryConfiguration.searchPathCompositeEnvironmentRepository
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:60) ~[spring-boot-autoconfigure-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.context.annotation.ConditionEvaluator.shouldSkip(ConditionEvaluator.java:108) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForBeanMethod(ConfigurationClassBeanDefinitionReader.java:181) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForConfigurationClass(ConfigurationClassBeanDefinitionReader.java:141) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitions(ConfigurationClassBeanDefinitionReader.java:117) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:327) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:232) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:275) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:95) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:705) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:531) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:141) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:744) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:391) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:312) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:120) ~[spring-boot-test-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:117) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:108) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:118) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.boot.test.autoconfigure.SpringBootDependencyInjectionTestExecutionListener.prepareTestInstance(SpringBootDependencyInjectionTestExecutionListener.java:43) ~[spring-boot-test-autoconfigure-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:246) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:227) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:289) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) ~[junit-4.12.jar:4.12]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:291) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:246) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) ~[junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) ~[junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) ~[junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) ~[junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) ~[junit-4.12.jar:4.12]
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363) ~[junit-4.12.jar:4.12]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190) ~[spring-test-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137) ~[junit-4.12.jar:4.12]
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68) ~[junit-rt.jar:na]
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33) ~[junit-rt.jar:na]
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230) ~[junit-rt.jar:na]
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58) ~[junit-rt.jar:na]
Caused by: java.lang.NullPointerException: null
	at java.base/java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:936) ~[na:na]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeanDefinition(DefaultListableBeanFactory.java:768) ~[spring-beans-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.cloud.config.server.composite.CompositeUtils.getEnvironmentRepositoryFactoryTypeParams(CompositeUtils.java:82) ~[spring-cloud-config-server-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.cloud.config.server.composite.OnSearchPathLocatorPresent.getMatchOutcome(OnSearchPathLocatorPresent.java:49) ~[spring-cloud-config-server-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:47) ~[spring-boot-autoconfigure-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	... 42 common frames omitted
```

--- a/spring-cloud-config-azure-keyvault-starter/src/main/java/org/srempfer/cloud/config/keyvault/autoconfigure/EnvironmentEncryptorAutoConfiguration.java
+++ b/spring-cloud-config-azure-keyvault-starter/src/main/java/org/srempfer/cloud/config/keyvault/autoconfigure/EnvironmentEncryptorAutoConfiguration.java
@@ -4,9 +4,7 @@ import com.microsoft.azure.keyvault.spring.KeyVaultOperation;
 import org.springframework.boot.autoconfigure.AutoConfigureAfter;
 import org.springframework.boot.autoconfigure.AutoConfigureBefore;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
-import org.springframework.cloud.config.server.config.ConfigServerMvcConfiguration;
-import org.springframework.cloud.config.server.config.EncryptionAutoConfiguration;
-import org.springframework.cloud.config.server.config.EnvironmentRepositoryConfiguration;
+import org.springframework.cloud.config.server.config.ConfigServerAutoConfiguration;
 import org.springframework.cloud.config.server.encryption.EnvironmentEncryptor;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
@@ -21,9 +19,9 @@ import java.util.List;
  *
  * @author Stefan Rempfer
  */
-@AutoConfigureBefore ( { EncryptionAutoConfiguration.class, ConfigServerMvcConfiguration.class, EnvironmentRepositoryConfiguration.class } )
+@AutoConfigureBefore ( { ConfigServerAutoConfiguration.class } )
 @AutoConfigureAfter ( KeyVaultAutoConfiguration.class )
-@Configuration
+@Configuration ( proxyBeanMethods = false )
 public class EnvironmentEncryptorAutoConfiguration {
 
     @Primary

--- a/spring-cloud-config-azure-keyvault-starter/src/main/java/org/srempfer/cloud/config/keyvault/autoconfigure/EnvironmentRepositoryAutoConfiguration.java
+++ b/spring-cloud-config-azure-keyvault-starter/src/main/java/org/srempfer/cloud/config/keyvault/autoconfigure/EnvironmentRepositoryAutoConfiguration.java
@@ -2,11 +2,14 @@ package org.srempfer.cloud.config.keyvault.autoconfigure;
 
 import com.microsoft.azure.keyvault.spring.KeyVaultOperation;
 import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.config.server.config.ConfigServerAutoConfiguration;
 import org.springframework.cloud.config.server.environment.EnvironmentRepository;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
 import org.springframework.context.annotation.Profile;
 import org.srempfer.cloud.config.keyvault.KeyVaultEnvironmentRepository;
 import org.srempfer.cloud.config.keyvault.KeyVaultEnvironmentRepositoryFactory;
@@ -16,18 +19,31 @@ import org.srempfer.cloud.config.keyvault.KeyVaultEnvironmentRepositoryFactory;
  *
  * @author Stefan Rempfer
  */
+@AutoConfigureBefore ( ConfigServerAutoConfiguration.class )
 @AutoConfigureAfter ( KeyVaultAutoConfiguration.class )
+@Import ( KeyVaultRepositoryConfiguration.class )
 @EnableConfigurationProperties ( KeyVaultEnvironmentProperties.class )
-@Configuration
+@Configuration ( proxyBeanMethods = false )
 public class EnvironmentRepositoryAutoConfiguration {
 
-    @ConditionalOnBean ( KeyVaultOperation.class )
-    @Bean
-    public KeyVaultEnvironmentRepositoryFactory keyVaultEnvironmentRepositoryFactory ( KeyVaultOperation keyVaultOperation ) {
-        return new KeyVaultEnvironmentRepositoryFactory ( keyVaultOperation );
-    }
+    @Configuration ( proxyBeanMethods = false )
+    static class KeyVaultFactoryConfig {
 
-    @Profile ( "keyvault" )
+        @ConditionalOnBean ( KeyVaultOperation.class )
+        // factory name have to start with the type name - see org.springframework.cloud.config.server.composite.CompositeUtils.getFactoryName()
+        // to align it with the name of the profile the name is explicitly defined with a lower 'v'
+        @Bean ( name = "keyvaultEnvironmentRepositoryFactory" )
+        public KeyVaultEnvironmentRepositoryFactory keyVaultEnvironmentRepositoryFactory ( KeyVaultOperation keyVaultOperation ) {
+            return new KeyVaultEnvironmentRepositoryFactory ( keyVaultOperation );
+        }
+
+    }
+}
+
+@Configuration ( proxyBeanMethods = false )
+@Profile ( "keyvault" )
+class KeyVaultRepositoryConfiguration {
+
     @ConditionalOnBean ( KeyVaultOperation.class )
     @Bean
     public KeyVaultEnvironmentRepository keyVaultEnvironmentRepository ( KeyVaultEnvironmentRepositoryFactory factory,

--- a/spring-cloud-config-azure-keyvault-starter/src/main/java/org/srempfer/cloud/config/keyvault/autoconfigure/KeyVaultAutoConfiguration.java
+++ b/spring-cloud-config-azure-keyvault-starter/src/main/java/org/srempfer/cloud/config/keyvault/autoconfigure/KeyVaultAutoConfiguration.java
@@ -25,7 +25,7 @@ import java.util.Collections;
  */
 @ConditionalOnProperty ( prefix = "spring.cloud.config.server.azure.keyvault", name = "uri" )
 @EnableConfigurationProperties ( KeyVaultProperties.class )
-@Configuration
+@Configuration ( proxyBeanMethods = false )
 public class KeyVaultAutoConfiguration {
 
     private static final Logger LOGGER = LoggerFactory.getLogger ( KeyVaultAutoConfiguration.class );

--- a/spring-cloud-config-azure-keyvault-starter/src/test/java/org/srempfer/cloud/config/keyvault/ConfigServerCompositeIntegrationTest.java
+++ b/spring-cloud-config-azure-keyvault-starter/src/test/java/org/srempfer/cloud/config/keyvault/ConfigServerCompositeIntegrationTest.java
@@ -1,6 +1,5 @@
 package org.srempfer.cloud.config.keyvault;
 
-import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
@@ -34,7 +33,6 @@ import static org.assertj.core.api.Assertions.entry;
         "spring.cloud.config.server.composite.[0].searchLocations=classpath:/config-data/,classpath:/config-data/{application}/",
         "spring.cloud.config.server.composite.[1].type=keyvault"
     } )
-@Ignore ( "Ignored because of issue #6" )
 public class ConfigServerCompositeIntegrationTest {
 
     @LocalServerPort
@@ -59,17 +57,17 @@ public class ConfigServerCompositeIntegrationTest {
         assertThat ( propertySources ).hasSize ( 2 );
 
         PropertySource propertySource = propertySources.get ( 0 );
-        assertThat ( propertySource.getName () ).isEqualTo ( "keyvault-application-default" );
-        Map<Object, Object> source = (Map<Object, Object>) propertySource.getSource ();
-        assertThat ( source ).contains ( entry ( "simplekey", "dummy" ) );
-
-        propertySource = propertySources.get ( 1 );
         assertThat ( propertySource.getName () ).isEqualTo ( "classpath:/config-data/application.properties" );
-        source = (Map<Object, Object>) propertySource.getSource ();
+        Map<Object, Object> source = (Map<Object, Object>) propertySource.getSource ();
         assertThat ( source ).containsOnly (
             entry ( "client.test.key", "test-value" ),
             entry ( "client.test.encrypted", "decrypted-value" ),
             entry ( "missing.client.test.encrypted-missing", "<n/a>" ) );
+
+        propertySource = propertySources.get ( 1 );
+        assertThat ( propertySource.getName () ).isEqualTo ( "keyvault-application-default" );
+        source = (Map<Object, Object>) propertySource.getSource ();
+        assertThat ( source ).contains ( entry ( "simplekey", "dummy" ) );
     }
 
     @Configuration

